packages/editor/src/editor/sidebar/fields/condition-builder/ConditionBuilder.tsx

```diff
@@ -0,0 +1,137 @@
+    const newGroups = [...conditionGroups];
+    newGroups[groupIndex].conditions.push({ argument1: '', operator: '==', argument2: '', logicalOperator: 'and' });
+    setConditionGroups(newGroups);
+  };
+
+  const removeCondition = (groupIndex: number, conditionIndex: number) => {
+    const newGroups = [...conditionGroups];
+    newGroups[groupIndex].conditions.splice(conditionIndex, 1);
+    setConditionGroups(newGroups);
+  };
+
+  const addConditionGroup = () => {
+    setConditionGroups([
+      ...conditionGroups,
+      { conditions: [{ argument1: '', operator: '==', argument2: '', logicalOperator: 'and' }], logicalOperator: 'and' }
+    ]);
+  };
+
+  const removeConditionGroup = (groupIndex: number) => {
+    const newGroups = [...conditionGroups];
+    newGroups.splice(groupIndex, 1);
+    setConditionGroups(newGroups);
+  };
+
+  const generateConditionString = () => {
+    const wrapInQuotesIfNeeded = (arg: string) => {
+      return arg.startsWith('data.') || arg === 'data' ? arg : `'${arg}'`;
+    };
+
+    const groupStrings = conditionGroups.map((group, index) => {
+      const conditions = group.conditions
+        .map((con, index) => {
+          const formattedArg1 = wrapInQuotesIfNeeded(con.argument1);
+          const formattedArg2 = wrapInQuotesIfNeeded(con.argument2);
+          const condition = `${formattedArg1} ${con.operator} ${formattedArg2}`;
+          const logicalOp = index < group.conditions.length - 1 ? ` ${con.logicalOperator.toLowerCase()} ` : '';
+          return con.operator === 'isTrue' ? formattedArg1 : condition + logicalOp;
+        })
+        .join('');
+      const logicGroupOp = index < conditionGroups.length - 1 ? ` ${group.logicalOperator.toLowerCase()} ` : '';
+      return logicGroupOp === '' ? (conditionGroups.length === 1 ? conditions : `(${conditions})`) : `(${conditions}) ${logicGroupOp} `;
+    });
+
+    return `#{${groupStrings.join('')}}`;
+  };
+
+  const handleAddCondition = () => {
+    const finalCondition = generateConditionString();
+    onChange(finalCondition);
+    apply();
+  };
+
+  return (
+    <Tabs value='Condition' style={{ height: '100%', overflow: 'auto' }}>
+      <Flex direction='column' gap={1} style={{ height: '100%' }}>
+        <TabsList>
+          <TabsTrigger value='Condition'>
+            <IvyIcon icon={IvyIcons.Attribute} />
+            Condition
+          </TabsTrigger>
+        </TabsList>
+        <Flex direction='column' gap={1} justifyContent='space-between' style={{ height: '100%', overflow: 'auto' }}>
+          <Flex direction='column' gap={2}>
+            <BasicCheckbox
+              label='Enable Grouping'
+              checked={isConditionGroupEnabled}
+              onCheckedChange={() => setIsConditionGroupEnabled(!isConditionGroupEnabled)}
+              disabled={conditionGroups.length > 1 && isConditionGroupEnabled}
+            />
+            {conditionGroups.map((group, groupIndex) => (
+              <ConditionGroup
```

Comment:
I’m not happy with the fact that I’m passing in so many functions and properties, but I couldn’t come up with a better solution quickly. Maybe it would be better to declare the respective functions directly within the component and only pass in conditionGroups and setConditionGroups, or even create a context or a useHook. What do you think? What would be the most elegant solution here?

Reply:
instead of create this methods here, can't you create where you need them and add the setConditions method as parameter?

Reply:
at least you should share the types of this functions

Reply:
I'm not quite sure what you mean by 'Share types of these functions' but I now only pass the setConditionGroups function into both ConditionGroup and Condition and create the necessary functions directly there 👍

Reply:
I mean that you reuse the types of the functions, e.g. reuse the types from Condition on the type ConditionGroup for some of the parameters.